src/clinical/clinical-api.ts

```diff
@@ -95,6 +95,59 @@ class ClinicalController {
     res.send(zip.toBuffer());
   }
 
+  @HasProgramReadAccess((req: Request) => req.params.programId)
+  async getSpecificClinicalDataAsTsvsInZip(req: Request, res: Response) {
+    const programId: string = req.params.programId;
+    const sort: string = 'donorId';
+    const page: number = 0;
+    const state: CompletionStates = req.query.completionState || CompletionStates.all;
+    const completionState: {} = completionFilters[state] || {};
+
+    const entityTypes: string[] =
+      req.query.entityTypes !== undefined ? req.query.entityTypes.split(',') : [];
+
+    const donorIds: number[] =
+      req.query.donorIds?.match(/\d*/gi)?.filter((match: string) => !!match && parseInt(match)) ||
+      [];
+
+    const submitterDonorIds: string[] =
+      req.query.submitterDonorIds && req.query.submitterDonorIds.length > 0
+        ? req.query.submitterDonorIds?.split(',').filter((match: string) => !!match)
+        : [];
+
+    const query: ClinicalQuery = {
+      ...req.query,
+      sort,
+      entityTypes,
+      page,
+      donorIds,
+      submitterDonorIds,
+      completionState,
+    };
+
+    if (!programId) {
+      return ControllerUtils.badRequest(res, 'Invalid programId provided');
+    }
+
+    const entityData = await service
+      .getPaginatedClinicalData(programId, query)
+      .then(data => data.clinicalEntities);
+
+    const todaysDate = currentDateFormatted();
+    res
+      .status(200)
+      .contentType('application/zip')
+      .attachment(`${programId}_Clinical_Data_${todaysDate}.zip`);
+
+    const zip = new AdmZip();
+    entityData.forEach((d: any) => {
+      const tsvData = TsvUtils.convertJsonRecordsToTsv(d.records, d.entityFields);
+      zip.addFile(`${d.entityName}.tsv`, Buffer.alloc(tsvData.length, tsvData));
+    });
```

Comment:
probably a utility method for building a zip of clinical data, or sometrhing, just trying not to repeat code blocks from the copy paste.

Reply:
I've added a `createClinicalZipFile` method which returns a zip file, take a look
It works specifically for ClinicalEntityData w/ a `records` property (may be reusable for other entity records w/ type update)